File: strategies/10-low-volatility-anomaly/strategy.py
"""Low-Volatility Anomaly strategy (time-series version).

Thesis: Baker, Bradley & Wurgler (2011) and Blitz & van Vliet (2007) document
that low-beta / low-volatility securities produce superior risk-adjusted returns
despite CAPM's prediction to the contrary. Two reinforcing mechanisms drive this:

  Behavioral (lottery demand): investors overpay for high-vol "lottery ticket"
  stocks, bidding up their prices and depressing their expected returns. Low-vol
  assets are overlooked, leaving their expected return intact or elevated.

  Institutional (leverage constraint): benchmark-constrained fund managers cannot
  use leverage to hit return targets, so they tilt toward high-vol stocks to boost
  raw expected returns without explicit leverage. This systematic demand overprices
  high-vol assets and underprices low-vol assets.

Cross-sectional to time-series adaptation: the published anomaly ranks stocks in a
universe and buys the lowest-vol decile. Adapted here to a single asset: the asset
is held when its OWN realized volatility is below its own rolling median (calm,
predictable, "low-vol anomaly" regime), and exited when vol spikes above the
rolling 75th percentile (high-vol, lottery-like regime). The economic rationale is
preserved: hold the asset during its calm periods, step aside during volatile ones.

Differs from strategy 09 (Volatility-Managed Portfolio): strategy 09 scales
position SIZE continuously by target_vol / realized_vol (always invested, just more
or less). This strategy makes a binary SELECTION decision (fully in or fully out)
based on RELATIVE volatility (vol vs own history), not absolute vol level. An asset
can pass strategy 09's size threshold while failing this strategy's selection
criterion if its current vol is high relative to its own past even if low in
absolute terms.

Differs from momentum strategies 02, 07, 08: those strategies go long during strong
recent price movement — often the highest-vol periods. This strategy enters in the
opposite regime: when the asset is calm and predictable.
"""

import numpy as np

DEFAULT_PARAMS = {"vol_window": 60, "ranking_window": 252, "exit_percentile": 75}


class LowVolatilityAnomaly:
    """Time-series Low-Volatility Anomaly strategy (Baker/Blitz entry criterion).

    At each bar t, computes the trailing vol_window-bar annualized realized
    volatility (std of daily pct_change returns × sqrt(252)). The strategy holds
    a long position while that volatility is below its own rolling median over the
    prior ranking_window bars, and exits when vol rises above the rolling
    exit_percentile-th percentile of the same window.

    Hysteresis prevents churn around the median boundary: once long, the position
    is held until vol explicitly exceeds the exit_percentile threshold, even if vol
    temporarily rises above the median during the hold.

    Entry condition: current_vol < rolling_median(vol_history[-ranking_window:])
    Exit condition:  current_vol > rolling_percentile(exit_percentile, same window)

    Warm-up: the strategy returns 0.0 for the first vol_window + ranking_window bars
    because no meaningful percentile can be computed before the full history is
    available.

    Args:
        vol_window: Lookback in bars for realized volatility. Default 60 (roughly one
            trading quarter). Must be at least 2.
        ranking_window: Lookback in bars for the rolling median / percentile of vol
            history. Default 252 (one trading year). Must be >= vol_window so that
            the percentile is computed over a meaningful vol series.
        exit_percentile: Percentile of the rolling vol history used as the exit
            threshold. Default 75. Must be in (0, 100).

    Raises:
        ValueError: If vol_window < 2, ranking_window < vol_window, or
            exit_percentile not in (0, 100).
    """

    def __init__(
        self,
        vol_window: int = 60,
        ranking_window: int = 252,
        exit_percentile: float = 75.0,
    ):
        if vol_window < 2:
            raise ValueError("vol_window must be at least 2")
        if ranking_window < vol_window:
            raise ValueError("ranking_window must be >= vol_window")
        if not (0 < exit_percentile < 100):
            raise ValueError("exit_percentile must be in (0, 100)")
        self.vol_window = vol_window
        self.ranking_window = ranking_window
        self.exit_percentile = exit_percentile
        self._in_position = False

    def __call__(self, view) -> float:
        """Compute entry/exit signal for the current bar.

        Requires at least vol_window + ranking_window bars of close history.
        During warm-up, returns 0.0 (flat).

        The realized vol series is computed as a rolling vol_window std of
        pct_change returns, annualized by sqrt(252). The last ranking_window
        values of that series form the vol history used for the median and
        exit_percentile thresholds.

        No look-ahead: all computations use close[t] and earlier. The engine
        fills at open[t+1].

        Args:
            view: A guarded DataFrame slice with OHLCV columns up to and
                including bar t.

        Returns:
            1.0 if long, 0.0 if flat.
        """
        closes = view["close"]
        n = len(closes)
        min_bars = self.vol_window + self.ranking_window
        if n < min_bars:
            self._in_position = False
            return 0.0

        returns = closes.pct_change()
        # Rolling annualized vol; iloc[-1] is current bar's vol estimate
        rolling_vol = returns.rolling(self.vol_window).std() * (252 ** 0.5)
        current_vol = float(rolling_vol.iloc[-1])

        if np.isnan(current_vol):
            self._in_position = False
            return 0.0

        # Use the last ranking_window values of the vol series as the reference
        vol_history = rolling_vol.dropna().iloc[-self.ranking_window :]
        if len(vol_history) < 2:
            self._in_position = False
            return 0.0

        rolling_median = float(vol_history.median())
        rolling_exit = float(np.percentile(vol_history.values, self.exit_percentile))

        if self._in_position:
            if current_vol > rolling_exit:
                self._in_position = False
                return 0.0
            return 1.0
        else:
            if current_vol < rolling_median:
                self._in_position = True
                return 1.0
            return 0.0

File: strategies/10-low-volatility-anomaly/strategy.ts
# Low-Volatility Anomaly — Baker, Bradley & Wurgler 2011 / Blitz & van Vliet 2007
# Thesis: the low-vol anomaly predicts that assets in a low-volatility regime
#   (vol below own historical median) produce superior risk-adjusted returns.
#   This time-series version goes long when current vol is below rolling median
#   and exits when vol rises above the exit_percentile threshold.
# Parameters:
#   volWindow=60      — bars for realized vol (one quarter). Prior-specified.
#   rankingWindow=252 — bars for rolling median/percentile (one year). Prior-specified.
#   exitPercentile=75 — percentile of rolling vol history used as exit threshold.
#                       Prior-specified from anomaly literature (75th = upper quartile).
# Vol proxy note: Python strategy.py uses returns-based realized vol
#   (pct_change().rolling(volWindow).std() * sqrt(252)). ThinkScript uses
#   StdDev(close, volWindow), which is price-level standard deviation, NOT a
#   returns-based measure. The two measures are monotonically related for a given
#   series over a rolling window — when returns-based vol rises, price-level std
#   also rises — so the signal direction (below-median = long; above-exit = flat)
#   is preserved even though the absolute scale differs. The README documents this
#   proxy substitution explicitly.
# Percentile approximation: thinkScript has no native rolling percentile function.
#   The exit threshold is approximated as:
#     Lowest(vol, rankingWindow) + exitPercentile/100 * (Highest - Lowest)
#   This is a min-max interpolated quantile, which approximates the true 75th
#   percentile when the vol distribution is roughly uniform or symmetric. For
#   skewed vol distributions the approximation may overestimate the true 75th
#   percentile; this makes the exit condition *less* sensitive (harder to exit)
#   than the Python strategy. The effect is a slightly longer hold duration; it
#   does not change the signal direction.
# Median approximation: thinkScript uses Average(vol, rankingWindow) (simple
#   moving average) as a proxy for the rolling median. For right-skewed vol
#   distributions (typical of realized volatility), the mean > median, so
#   the average-based entry threshold is slightly harder to cross (higher bar)
#   than the Python median. The practical effect is marginally fewer entries.
# Signal logic: addOrder BUY_AUTO when in low-vol regime; SELL_TO_CLOSE on exit.
# Load: Charts > Studies > Edit Studies > Strategies tab > Create > paste > apply Daily.
# Fill note: open[-1] = next-bar open, matching Python engine t+1 open fill model.

input volWindow = 60;
input rankingWindow = 252;
input exitPercentile = 75;

# Realized vol proxy: price-level std dev over volWindow bars (see note above)
def currentVol = StdDev(close, volWindow);

# Rolling vol history statistics over rankingWindow bars
def volLow = Lowest(currentVol, rankingWindow);
def volHigh = Highest(currentVol, rankingWindow);

# Median approximation via simple moving average (see note above)
def volMedian = Average(currentVol, rankingWindow);

# Exit threshold via min-max interpolation (see note above)
def volExit = volLow + (exitPercentile / 100.0) * (volHigh - volLow);

# Warm-up guard: no signal until enough history is available
def warmedUp = BarNumber() > (volWindow + rankingWindow);

# Long entry: current vol is in the low-vol regime (below rolling median)
AddOrder(OrderType.BUY_AUTO,
         warmedUp and currentVol < volMedian,
         open[-1], 1,
         Color.GREEN, Color.GREEN, "LowVol Entry");

# Exit: vol has risen into the high-vol regime (above exit threshold)
AddOrder(OrderType.SELL_TO_CLOSE,
         warmedUp and currentVol > volExit,
         open[-1], 1,
         Color.RED, Color.RED, "LowVol Exit");

# Vol overlay for visualization
plot volLine = if warmedUp then currentVol else Double.NaN;
volLine.SetDefaultColor(Color.YELLOW);
volLine.SetLineWeight(1);

# Median reference line
plot medianLine = if warmedUp then volMedian else Double.NaN;
medianLine.SetDefaultColor(Color.CYAN);
medianLine.SetStyle(Curve.SHORT_DASH);

# Exit threshold reference line
plot exitLine = if warmedUp then volExit else Double.NaN;
exitLine.SetDefaultColor(Color.RED);
exitLine.SetStyle(Curve.SHORT_DASH);
